webdriver_dotnet/SemanticLocators/SemanticLocators.Tests/TestBase.cs

```diff
@@ -0,0 +1,188 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SemanticLocatorException.cs" company="The Semantic Locators Authors">
+//
+// Copyright 2021 The Semantic Locators Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// </copyright>
+//-----------------------------------------------------------------------
+using NUnit.Framework;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Chrome;
+using OpenQA.Selenium.Firefox;
+using OpenQA.Selenium.Remote;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SemanticLocators.Tests
+{
+    public class TestBase
+    {
+        // Use a string to identify browsers so the test names are readable
+        private static ConcurrentDictionary<string, IWebDriver> Drivers = new ConcurrentDictionary<string, IWebDriver>();
+        private static string[] _supportedBrowsers = new[] { "chrome", "firefox" };
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            foreach (var driver in Drivers)
+                driver.Value.Quit();
+        }
+
+        protected void RenderHtml(string html, IWebDriver driver)
+        {
+            string browserName = (string)((RemoteWebDriver)driver).Capabilities.GetCapability("browserName");
+            // IE doesn't support data URLs
+            if (browserName.Equals("internet explorer"))
+            {
+                driver.Navigate().GoToUrl("about:blank");
+                ((IJavaScriptExecutor)driver)
+                    .ExecuteScript(
+                        "document.write(\""
+                         + html.Replace("\\", "\\\\") // Extra escape for the JS string
+                         + "\")");
+
+            }
+            else
+            {
+                driver.Navigate().GoToUrl("data:text/html;charset=utf-8," + html);
+            }
+        }
+
+        protected static IWebDriver GetDriver(string driverName)
+        {
+            if (Drivers.ContainsKey(TestContext.CurrentContext.Test.ID))
```

Comment:
Currently the tests take ~1m30s to run on my machine. This is because:
- Using the test ID as the key to `Drivers` means a new WebDriver instance is brought up for each test. Please use `driverName` as the key to `Drivers` so they can be reused
- Running the tests in parallel also prevents them from sharing drivers.

Removing the parallelism and using `driverName` for the key makes the suite run much faster, in 20s for me.

Reply:
Very interesting, on my Windows machine I was getting an almost opposite result, hence the parallelism